File: node/src/catalog/registry.ts
import type { InteractionSummary, ParticipantSummary } from '../types.js';

export interface InteractionMetadata {
  id: string;
  workstreamId: string;
  workstreamName?: string;
  name?: string;
  description?: string;
  status?: string;
  sourceParticipantId: string;
  targetParticipantId: string;
  sourceDid?: string;
  targetDid?: string;
  type?: string;
  actor?: string;
  states?: string[];
  roiClassification?: string;
  roiCost?: number;
  roiTime?: number;
}

export interface ParticipantMetadata {
  id: string;
  did: string;
  name?: string;
  status?: string;
  customerId?: string;
  workstreamId?: string;
  workstreamName?: string;
}

/**
 * In-memory cache for interaction and participant metadata.
 */
export class Registry {
  private readonly interactions = new Map<string, InteractionMetadata>();
  private readonly participants = new Map<string, ParticipantMetadata>();

  replaceInteractions(items: InteractionMetadata[]): void {
    this.interactions.clear();
    for (const item of items) {
      if (!item.id) {
        continue;
      }
      this.interactions.set(item.id, {
        ...item,
        states: item.states ? [...item.states] : undefined
      });
    }
  }

  replaceParticipants(items: ParticipantMetadata[]): void {
    this.participants.clear();
    for (const item of items) {
      if (!item.id) {
        continue;
      }
      this.participants.set(item.id, { ...item });
    }
  }

  interaction(id: string): InteractionMetadata | undefined {
    const metadata = this.interactions.get(id);
    return metadata
      ? {
          ...metadata,
          states: metadata.states ? [...metadata.states] : undefined
        }
      : undefined;
  }

  interactionsList(): InteractionSummary[] {
    return Array.from(this.interactions.values()).map((item) => ({
      id: item.id,
      workstreamId: item.workstreamId,
      channelId: item.workstreamId,
      workstreamName: item.workstreamName,
      name: item.name,
      description: item.description,
      status: item.status,
      sourceParticipantId: item.sourceParticipantId,
      targetParticipantId: item.targetParticipantId,
      sourceDid: item.sourceDid,
      targetDid: item.targetDid,
      type: item.type,
      actor: item.actor,
      states: item.states ? [...item.states] : undefined,
      roiClassification: item.roiClassification,
      roiCost: item.roiCost,
      roiTime: item.roiTime
    }));
  }

  participantsList(): ParticipantSummary[] {
    return Array.from(this.participants.values()).map((item) => ({
      id: item.id,
      did: item.did,
      name: item.name,
      status: item.status,
      customerId: item.customerId,
      workstreamId: item.workstreamId,
      channelId: item.workstreamId,
      workstreamName: item.workstreamName
    }));
  }
}

File: go/internal/catalog/registry.go
package catalog

import "sync"

// InteractionType describes whether an interaction represents a touch, transition, or transfer event.
type InteractionType string

const (
	InteractionTypeTouch      InteractionType = "touch"
	InteractionTypeTransition InteractionType = "transition"
	InteractionTypeTransfer   InteractionType = "transfer"
)

// InteractionActor describes the agent responsible for an interaction.
type InteractionActor string

const (
	InteractionActorHuman  InteractionActor = "human"
	InteractionActorAI     InteractionActor = "ai"
	InteractionActorHybrid InteractionActor = "hybrid"
	InteractionActorNonAI  InteractionActor = "non-ai"
)

// ROIClassification describes how an interaction affects ROI metrics.
type ROIClassification string

const (
	ROIClassificationBaseline  ROIClassification = "baseline"
	ROIClassificationIncrement ROIClassification = "increment"
	ROIClassificationSavings   ROIClassification = "savings"
)

// Interaction holds the subsets of interaction metadata cached by the SDK.
type Interaction struct {
	ID                  string
	WorkstreamID        string
	SourceParticipantID string
	TargetParticipantID string
	SourceDID           string
	TargetDID           string
	Type                InteractionType
	Actor               InteractionActor
	States              []string
	ROIClassification   ROIClassification
	ROICost             int
	ROITime             int
}

// Participant captures an ID to DID mapping for cached directory lookups.
type Participant struct {
	ID           string
	DID          string
	Name         string
	Status       string
	CustomerID   string
	WorkstreamID string
}

// Registry is a threadsafe cache for interaction and participant directories.
type Registry struct {
	mu           sync.RWMutex
	interactions map[string]Interaction
	participants map[string]Participant
}

// NewRegistry constructs an empty registry instance.
func NewRegistry() *Registry {
	return &Registry{
		interactions: make(map[string]Interaction),
		participants: make(map[string]Participant),
	}
}

// ReplaceInteractions swaps the cached interaction metadata for the provided set.
func (r *Registry) ReplaceInteractions(items []Interaction) {
	snapshot := make(map[string]Interaction, len(items))
	for _, item := range items {
		if item.ID == "" {
			continue
		}
		snapshot[item.ID] = item
	}

	r.mu.Lock()
	r.interactions = snapshot
	r.mu.Unlock()
}

// Interaction returns the cached metadata for the supplied interaction ID.
func (r *Registry) Interaction(id string) (Interaction, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	item, ok := r.interactions[id]
	return item, ok
}

// Interactions returns a snapshot of all cached interaction records.
func (r *Registry) Interactions() []Interaction {
	r.mu.RLock()
	defer r.mu.RUnlock()

	result := make([]Interaction, 0, len(r.interactions))
	for _, item := range r.interactions {
		result = append(result, item)
	}
	return result
}

// ReplaceParticipants swaps the cached participant directory for the provided set.
func (r *Registry) ReplaceParticipants(items []Participant) {
	snapshot := make(map[string]Participant, len(items))
	for _, item := range items {
		if item.ID == "" || item.DID == "" {
			continue
		}
		snapshot[item.ID] = item
	}

	r.mu.Lock()
	r.participants = snapshot
	r.mu.Unlock()
}

// Participants returns a snapshot of all cached participant records.
func (r *Registry) Participants() []Participant {
	r.mu.RLock()
	defer r.mu.RUnlock()

	result := make([]Participant, 0, len(r.participants))
	for _, item := range r.participants {
		result = append(result, item)
	}
	return result
}

// ParticipantDID returns the DID for a participant ID, if present in the cache.
func (r *Registry) ParticipantDID(id string) (string, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	item, ok := r.participants[id]
	if !ok {
		return "", false
	}
	return item.DID, true
}
